Mathlib/Topology/GPseudoMetric/Basic.lean

```diff
@@ -0,0 +1,328 @@
+Authors: Edward van de Meent
+-/
+import Mathlib.Algebra.CovariantAndContravariant
+import Mathlib.Algebra.NeZero
+import Mathlib.Algebra.Order.Monoid.Lemmas
+import Mathlib.Data.Set.Lattice
+/-!
+# General Pseudo-Metrics
+
+In this file we introduce `GPseudoMetric`.
+This differs from `PseudoMetricSpace` by not requiring that the codomain of the metric be `ℝ`,
+instead only requiring that it is an (additive) commutative monoid, with a linear ordering.
+
+## Main Definitions
+
+- `GPseudoMetric α β`: A structure containing a distance function on `α` with codomain `β`,
+which may be equal to 0 for non-equal elements. the distance function is 0 for equal elements,
+is commutative in its arguments, and satisifies the triangle inequality.
+- `GPseudoMetricClass α β`: The class of types of generic pseudo metrics on `α` to `β`.
+
+Additional useful definitions:
+
+- `ball gdist x δ`: The set of points with distance to x strictly less than δ
+- `closedBall gdist x δ`: The set of points with distance to x less than or equal to δ
+- `sphere gdist x δ`: The set of points with distance to x equal to δ
+
+-/
+
+open Set
+
+/-- Generic pseudo metrics
+
+A generic pseudo metric is a distance function `gdist : α → α → β`, which is zero for
+identical elements, for which the arguments commute, and for which the triangle
+inequality holds. As opposed to a classical pseudo metric, the codomain of this distance
+function is not necessarily ℝ (or ℝ≥0∞), and as a result does not endow α with a uniform
+space.
+-/
+structure GPseudoMetric
+    (α : Type*) (β : Type*) [LinearOrder β] [AddCommMonoid β]
+    where
+  /-- A distance function on α with values in β -/
+  toFun (x y : α):β
+  gdist_self : ∀ x : α, toFun x x = 0
+  comm' : ∀ x y : α, toFun x y = toFun y x
+  triangle' : ∀ x y z : α, toFun x z ≤ toFun x y + toFun y z
+
+variable {α β :Type*} [LinearOrder β] [AddCommMonoid β] [CovariantClass β β (.+.) (.≤.)]
+@[ext]
+theorem GPseudoMetric.ext {d₁ d₂ : GPseudoMetric α β} (h : d₁.toFun = d₂.toFun) : d₁ = d₂ := by
+  cases' d₁ with d _ _ _ B hB
+  cases' d₂ with d' _ _ _ B' hB'
+  obtain rfl : d = d' := h
+  congr
+
+instance : FunLike (GPseudoMetric α β) α (α → β) where
+  coe := GPseudoMetric.toFun
+  coe_injective' := by apply GPseudoMetric.ext
+
+/-- A class for types of pseudo metric functions on α with values in β -/
+class GPseudoMetricClass (T : Type*) (α β : outParam Type*) [LinearOrder β] [AddCommMonoid β]
+    [CovariantClass β β (.+.) (.≤.)] [FunLike T α (α → β)] : Prop :=
+  gdist_self : ∀ (gdist:T), ∀ x: α, gdist x x = 0
+  comm' : ∀ (gdist:T), ∀ x y:α, gdist x y = gdist y x
+  triangle' : ∀ (gdist:T), ∀ (x y z:α), gdist x z ≤ gdist x y + gdist y z
+
+instance: GPseudoMetricClass (GPseudoMetric α β) α β where
+  gdist_self := GPseudoMetric.gdist_self
+  comm' := GPseudoMetric.comm'
+  triangle' := GPseudoMetric.triangle'
+
+
+variable {T:Type*} [FunLike T α (α → β)] [GPseudoMetricClass T α β] (gdist:T)
+
+
```

Comment:
As once above, please add spaces around the colon - and one blank line around them is enough :-)
```suggestion
variable {T : Type*} [FunLike T α (α → β)] [GPseudoMetricClass T α β] (gdist : T)